osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs

```diff
@@ -230,6 +273,34 @@ private PathType convertPathType(string input)
             }
         }
 
+        private struct PointOrPathType
```

Comment:
With all due respect, as bad as the [previous double indexing was](https://github.com/ppy/osu/pull/23282#discussion_r1176956392), I think this is worse. There isn't a world where an almost-C-style union with an "or" in the name makes *any* problem better than literally any other solution.

I am in favour of reverting https://github.com/ppy/osu/pull/23282/commits/4662504d6e91da2fbbc0e873326e2b9a31afe300.

Reply:
Honestly, it's clear at my point of view to first convert the string (span) into tokens, then parse from the token stream. The union represents tokens in the source string.

Do you think it's better to convert the input string into points, the range of points (instead of string segmentations) of each segment, and types of each segment? This can also eliminate the `endPoint` and make it naturally a member of both segments.

Reply:
The `endPoint` that lives beyond the next type controller is really annoying. Otherwise the input can be scanned sequentially in total. 

> Let's not overcomplicate and refactor things in what is a mostly mechanical pull.

For sure. I'll revert it to closet state, and refactor it in a follow-up PR.